File: 03-stream-processing/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise4/BackpressureJob.scala
package com.becomedataengineer.exercise4

import org.apache.spark.sql.catalyst.ScalaReflection
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{Dataset, SparkSession, functions}

import java.sql.Timestamp

object DeadLetterQueueJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._

    val inputStream = sparkSession.readStream.format("kafka")
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:29092",
        "subscribe" -> "data_for_backpressure",
        "startingOffsets" -> "EARLIEST"
      )).load()


    val rowToProcessSchema = ScalaReflection.schemaFor[RowToProcess].dataType.asInstanceOf[StructType]

    val rowsToProcess = inputStream
      .select(functions.from_json($"value".cast("string"), rowToProcessSchema).as("value"))
      .selectExpr("value.*")
      .as[RowToProcess]

    val writeQuery = rowsToProcess.writeStream
      .foreachBatch((realTimeDataset: Dataset[RowToProcess], batchNumber: Long) => {
        val maybeBackpressuredDataset = if (batchNumber > 0) {
          Some(sparkSession.read.schema(rowToProcessSchema).json(s"${BackpressureDir}/${batchNumber - 1}")
          .as[RowToProcess])
        } else {
          None
        }
        val datasetToProcess = maybeBackpressuredDataset
          .map(backpressuredDataset => backpressuredDataset.union(realTimeDataset))
          .getOrElse(realTimeDataset)

        // Classify the records for the backpressure storage
        val maxRecordsPerJob = 5
        val backpressureWindowSpec = Window.partitionBy("id").orderBy($"generation_time".asc)
        val datasetWithBackpressureFlag = datasetToProcess.withColumn("is_backpressured", functions.when(
          functions.row_number().over(backpressureWindowSpec) > maxRecordsPerJob, true
        ).otherwise(false))

        val cachedDatasetWithBackpressureFlag = datasetWithBackpressureFlag.cache()

        cachedDatasetWithBackpressureFlag.filter("is_backpressured = true")
          .drop("is_backpressured")
          .write.format("json").save(s"${BackpressureDir}/${batchNumber}")

        cachedDatasetWithBackpressureFlag.filter("is_backpressured = false")
          .drop("is_backpressured")
          .show(truncate = false)

        cachedDatasetWithBackpressureFlag.unpersist()
        ()
      })
      .start()

    writeQuery.awaitTermination()
  }

}

case class RowToProcess(id: Int, row_value: String, generation_time: Timestamp)

File: 03-stream-processing/solution-proposal/python/exercise4/backpressure_job.py
import shutil

from pyspark.sql import SparkSession, DataFrame, Window
from pyspark.sql import functions as F
from pyspark.sql.types import StructField, TimestampType, IntegerType, StringType, StructType

backpressure_dir = '/tmp/bde/module5/homework/e4/backpressure'
shutil.rmtree(backpressure_dir, ignore_errors=True)

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
    .getOrCreate()

input_data_stream = spark.readStream \
    .option("kafka.bootstrap.servers", "localhost:29092") \
    .option("subscribe", "data_for_backpressure") \
    .option("startingOffsets", "EARLIEST") \
    .format("kafka") \
    .load()

json_schema = StructType([
    StructField("id", IntegerType()), StructField("row_value", StringType()),
    StructField("generation_time", TimestampType())
])

rows_to_process = input_data_stream.withColumn("filtering_struct",
                                                 F.from_json(F.col("value").cast("string"), json_schema))\
    .selectExpr("filtering_struct.*")

max_records_per_job = 5


def print_data_with_backpressure_mechanism(data_frame: DataFrame, batch_version: int) -> None:
    backpressure_dataset = None
    if batch_version > 0:
        backpressure_dataset = spark.read.schema(json_schema).json(f'{backpressure_dir}/{batch_version - 1}')

    dataset_to_process = data_frame if not backpressure_dataset else backpressure_dataset.union(data_frame)

    # Classify the records for the backpressure storage
    backpressure_window_spec = Window.partitionBy("id").orderBy(F.asc("generation_time"))
    dataset_to_process_with_backpressure_flag = dataset_to_process \
        .withColumn("is_backpressured", F.when(
        F.row_number().over(backpressure_window_spec) > max_records_per_job, True).otherwise(False))

    cached_dataset_with_backpressure_flag = dataset_to_process_with_backpressure_flag.cache()

    cached_dataset_with_backpressure_flag.filter("is_backpressured = true").drop("is_backpressured").write.format("json") \
        .mode('overwrite').save(f'{backpressure_dir}/{batch_version}')

    cached_dataset_with_backpressure_flag.filter("is_backpressured = false").drop("is_backpressured").show(truncate=False)

    cached_dataset_with_backpressure_flag.unpersist()
    return None


output_stream = rows_to_process.writeStream.foreachBatch(print_data_with_backpressure_mechanism)

output_stream.start().awaitTermination()
